haddock-api/src/Haddock/Backends/Hyperlinker/Parser.hs

```diff
@@ -0,0 +1,235 @@
+module Haddock.Backends.Hyperlinker.Parser
+    ( parse
+    , Token(..), TokenType(..)
+    , Position(..), Span(..)
+    ) where
+
+import Data.Char
+import Data.List
+import Data.Maybe
+
+data Token = Token
+    { tkType :: TokenType
+    , tkValue :: String
+    , tkSpan :: Span
+    }
+
+data Position = Position
+    { posRow :: !Int
+    , posCol :: !Int
+    }
+
+data Span = Span
+    { spStart :: Position
+    , spEnd :: Position
+    }
+
+data TokenType
+    = TkIdentifier
+    | TkKeyword
+    | TkString
+    | TkChar
+    | TkNumber
+    | TkOperator
+    | TkGlyph
+    | TkSpecial
+    | TkSpace
+    | TkComment
+    | TkCpp
+    | TkPragma
+    | TkUnknown
+    deriving (Show, Eq)
+
+-- | Turn source code string into a stream of more descriptive tokens.
+--
+-- Result should retain original file layout (including comments, whitespace,
+-- etc.), i.e. the following "law" should hold:
+--
+-- @concat . map 'tkValue' . 'parse' = id@
```

Comment:
Haddock actually supports `prop>` markup which when combined with `doctest` (and `QuickCheck`) can verify this this law in test phase. Might be worthwhile though not necessarily that high priority, I'm expecting there's a test for it already (haven't read down that far yet). It's similar to `>>>` in that it's code that can be ran as part of a testsuite.